protostar/cheatable_starknet/controllers/expect_call_controller.py

```diff
@@ -0,0 +1,96 @@
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from protostar.testing.test_environment_exceptions import ExpectedCallException
+from protostar.starknet import CairoOrPythonData, Address
+from protostar.starknet.selector import Selector
+
+if TYPE_CHECKING:
+    from protostar.testing import Hook
+    from protostar.cheatable_starknet.cheatables.cheatable_cached_state import (
+        CheatableCachedState,
+    )
+
+
+@dataclass
+class ExpectedCall:
+    address: Address
+    fn_selector: Selector
+    calldata: CairoOrPythonData
+
+
+def remove_expected_call_external(
+    expected_call: ExpectedCall, cheatable_state: "CheatableCachedState"
+):
+    data_for_address = cheatable_state.expected_contract_calls.get(
+        expected_call.address
+    )
+    if data_for_address is not None:
+        for index, (selector, calldata) in enumerate(data_for_address):
```

Comment:
it could be implemented as a `__eq__` method for `ExpectedCall` dataclass

Reply:
If we could have like a `FunctionCall` (name up to you) class which contains `selector` and `calldata`, we could store it in a map as a value.

Then the structure would look like this:
```
@dataclass
class FunctionCall:
    fn_selector: Selector
    calldata: CairoOrPythonData
    def __eq__(self, other):
        ... # Compare selector & calldata

@dataclass
class ExpectedCall:
    address: Address
    call: FunctionCall
    def __eq__(self, other):
        ... # Compare call and address
```

And you can then iterate over expected calls for address and compare with a `==` operator.

Also: please consider renaming `data_for_address` since it does not indicate what the data actually is. My proposition: `expected_calls` (leaving final decision up to you).

Reply:
good idea 👍

Reply:
[improve types for expected calls](https://github.com/software-mansion/protostar/pull/1481/commits/05258cda82feed9101e5989c605328b565091a41)